lib/galaxy/workflow/modules.py

```
@@ -381,6 +382,39 @@ def get_runtime_state(self):
         state.inputs = dict()
         return state
 
+    def get_runtime_inputs(self, connections=None):
+        print(self.subworkflow)
+        inputs = {}
+        for step in self.subworkflow.steps:
+            if step.type == "tool":
+                tool = step.module.tool
+                tool_inputs = step.module.state
+
+                def callback(input, prefixed_name, prefixed_label, value=None, **kwds):
+                    if input.type in ['data', 'data_collection']:
+                        return
```

Don't we want to raise an exception here and prevent running these workflows ?

---

I think data parameters with input connections are represented in the state as RuntimeValues still at this point - I guess because the data parameter is filled in later in the process. I had to put this check in for a perfectly valid workflow otherwise. This deserves a comment at least though for sure.

---

Ah, that makes sense. We can definitely address this at a later point if you prefer. I do think it'd be nice to error out ASAP though for the user.

---

I'm confused - this isn't an error - perfectly fine connected data inputs that are fine are represented as RuntimeValues. And the UI doesn't allow data parameters to be specified as RuntimeValues.

I guess you're suggesting we should produce proper errors on disconnected inputs - this I agree with but probably that should happen somewhere else? I suspect if we just stuck the disconnected data inputs in here they would work but I hate that a lot.

---

I was more thinking that for runtime data inputs inside subworkflows we'll want to raise an error, but maybe I'm misunderstanding what's going on here.

---

Nevermind, I think I misunderstood this.

---

> I guess you're suggesting we should produce proper errors on disconnected inputs - this I agree with but probably that should happen somewhere else?

Right, I thought that would be a good place, but I see now that we don't know which inputs are disconnected at this point.

> I suspect if we just stuck the disconnected data inputs in here they would work but I hate that a lot.

I was curious and tried that out, yes it does, but I agree that's not a good idea.